### ISSUE TITLE
Inconsistant API for `onEdit`

### ISSUE THREAD

**Issue by eric-burel** [VulcanJS/Vulcan#1883]
Hi,
Following a discussion with @justinr1234 on Slack, I'd like to point that the `onEdit` callback on schemas have an inconsistant API with `onInsert` and `onRemove`:

```js
onInsert(document, currentUser)
onEdit(modifier, document, currentUser)
onRemove(document, currentUser)
```
`onEdit` needs a `modifier` argument, the issue is that it is the first argument of the callback, and not the last as you would expect. One can easily be tricked by this (starting with me ^^), and think that the `onEdit` signature is `onEdit(document, currentUser, modifier)` which is a bit more natural.
Then Mongo will go nut because you added a field at the root of the `modifier` believing you were working on the `document` argument. The error message will then be cryptic Mongo error, that is, as any of those lovely Mongo error, hard to debug and completely undocumented in their doc. The issue is easy to fix but hard to find.

So the solutions could be:
1) change the API. This is breaking and error-prone for people used to Vulcan, so if it has to be done it could wait the 2.0 version maybe
2) prevent invalid modifier, maybe throwing a warning. Might tedious though and breakable, because Mongo API might evolve.
3) provide less cryptic error messages. Same as option 2), this might be tedious. I don't know if it's Vulcan job to enhance cryptic Mongo/Meteor errors, that's also a lot of responsibility.
4) Dismiss the issue, it is not an emergency at all and can be fixed by reading the doc carefully

**Comment by SachaG** [comment]
Maybe a better API would be `onEdit({ modifier, document, currentUser })`? this way the order wouldn't matter?

**Comment by eric-burel** [comment]
Yeah maybe an object as arg would be good. It makes the API a bit heavier but safer, and extensible if you'd like to add other objects to those callbacks like the context.

You can not fully rely on the number of argument, because you'll sometime write `onEdit(document)` (e.g when you want to create a field that will work as a index from other fields). I guess simply adding a deprecation message e.g `onEdit, onInsert, onRemove now expects an object as argument` would also help.
